pkg/destroy/vsphere/virtualmachines.go

```diff
@@ -0,0 +1,128 @@
+package vsphere
+
+import (
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+func (o *ClusterUninstaller) getVirtualMachineManagedObjects(moRef []types.ManagedObjectReference) ([]mo.VirtualMachine, error) {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	var virtualMachineMoList []mo.VirtualMachine
+	pc := property.DefaultCollector(o.Client)
+	err := pc.Retrieve(ctx, moRef, nil, &virtualMachineMoList)
+	if err != nil {
+		return nil, err
+	}
+	return virtualMachineMoList, nil
+}
+
+func (o *ClusterUninstaller) listVirtualMachines() ([]mo.VirtualMachine, error) {
+	tagAttachedObjects, err := o.getAttachedObjectsOnTag()
+	if err != nil && !isNotFound(err) {
+		return nil, err
+	}
+
+	// Seperate the objects attached to the tag based on type
+	// We only need VirtualMachine
+	var virtualMachineList []types.ManagedObjectReference
+	for _, attachedObject := range tagAttachedObjects {
+		for _, ref := range attachedObject.ObjectIDs {
+			if ref.Reference().Type == "VirtualMachine" {
+				virtualMachineList = append(virtualMachineList, ref.Reference())
+			}
+		}
+	}
+
+	var virtualMachineMoList []mo.VirtualMachine
+	if len(virtualMachineList) > 0 {
+		virtualMachineMoList, err = o.getVirtualMachineManagedObjects(virtualMachineList)
+	}
+
+	return virtualMachineMoList, err
+}
+
+func isPoweredOff(vmMO mo.VirtualMachine) bool {
+	return vmMO.Summary.Runtime.PowerState == "poweredOff"
+}
+
+func (o *ClusterUninstaller) stopVirtualMachine(vmMO mo.VirtualMachine) error {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	virtualMachineLogger := o.Logger.WithField("VirtualMachine", vmMO.Name)
+	if !isPoweredOff(vmMO) {
+		vm := object.NewVirtualMachine(o.Client, vmMO.Reference())
+		task, err := vm.PowerOff(ctx)
+		if err == nil {
+			err = task.Wait(ctx)
```

Comment:
Similar question here to the question about calling `Destroy`. Does calling `PowerOff` wait for the VM to actually power off before returning? If not, then we need some subsequent wait for the VM to be reporting that it is actually powered off before proceeding to the stage where we are destroying the VM.

Reply:
`PowerOff` [1] returns a `Task` [2] from a soap roundtrip request. There is `WaitForPowerState` [3] in the API but I haven't seen that one used anywhere. I'm new to the `vsphere` API, if you know of a better way, let me know and I'll change the code.

[1]  https://github.com/vmware/govmomi/blob/master/object/virtual_machine.go#L113-L124
[2] https://github.com/vmware/govmomi/blob/master/vim25/methods/methods.go#L9114-L9124
[3] https://github.com/vmware/govmomi/blob/master/object/virtual_machine.go#L819-L840